Source/WTF/wtf/FunctionPtr.h

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <memory>
+#include <wtf/FastMalloc.h>
+#include <wtf/Forward.h>
+#include <wtf/FunctionTraits.h>
+#include <wtf/Hasher.h>
+#include <wtf/PtrTag.h>
+
+namespace WTF {
+
+#if CPU(ARM_THUMB2) && ENABLE(JIT)
+// ARM instructions must be 16-bit aligned. Thumb2 code pointers to be loaded into
+// into the processor are decorated with the bottom bit set, while traditional ARM has
+// the lower bit clear. Since we don't know what kind of pointer, we check for both
+// decorated and undecorated null.
+#define ASSERT_NULL_OR_VALID_CODE_POINTER(ptr) \
+    ASSERT(!ptr || reinterpret_cast<intptr_t>(ptr) & ~1)
+#else
+#define ASSERT_NULL_OR_VALID_CODE_POINTER(ptr) // Anything goes!
+#endif
+
+template<PtrTag tag, typename Out, typename... In>
+class TypedFunctionPtr {
+public:
+    using Ptr = Out(*)(In...);
+
+    constexpr TypedFunctionPtr() = default;
+    constexpr TypedFunctionPtr(std::nullptr_t) : m_ptr(nullptr) { }
+
+    constexpr TypedFunctionPtr(Out(*ptr)(In...))
+        : m_ptr(encode(ptr))
+    { }
+
+// MSVC doesn't seem to treat functions with different calling conventions as
+// different types; these methods already defined for fastcall, below.
+#if CALLING_CONVENTION_IS_STDCALL && !OS(WINDOWS)
+    constexpr TypedFunctionPtr(Out(CDECL *ptr)(In...))
+        : m_ptr(encode(ptr))
+    { }
+#endif
+
+#if COMPILER_SUPPORTS(FASTCALL_CALLING_CONVENTION)
+    constexpr TypedFunctionPtr(Out(FASTCALL *ptr)(In...))
+        : m_ptr(encode(ptr))
+    { }
+#endif
+
+    Out operator()(In... in) const
```

Comment:
`Out operator()(In&&... in) const` doesn't work.  class `Function` also doe not use it.  I still haven't figured out why yet.  The error message implies that it's expecting to forward just one argument rather than all the arguments.  Will look more later.